File: src/lib/diet-phase.test.ts
import { describe, it, expect } from "vitest";
import {
  assessDietPhase,
  computeLeanMassSignal,
  computeWeightTrend,
  dietPhaseLabel,
  type DietPhaseInput,
} from "./diet-phase";

/** Weigh-ins every `stepDays` days, changing by `deltaPerWeek` lbs per week. */
function series(
  startWeight: number,
  weeks: number,
  deltaPerWeek: number,
  stepDays = 2,
  bodyFatStart?: number,
  bodyFatEnd?: number,
): { date: string; weight: number; bodyFatPercent?: number }[] {
  const out: { date: string; weight: number; bodyFatPercent?: number }[] = [];
  const totalDays = weeks * 7;
  const start = new Date("2026-05-01T12:00:00");

  for (let day = 0; day <= totalDays; day += stepDays) {
    const d = new Date(start);
    d.setDate(d.getDate() + day);
    const weight = startWeight + (deltaPerWeek * day) / 7;
    const point: { date: string; weight: number; bodyFatPercent?: number } = {
      date: d.toISOString().slice(0, 10),
      weight: Math.round(weight * 10) / 10,
    };
    if (bodyFatStart != null && bodyFatEnd != null) {
      point.bodyFatPercent =
        Math.round((bodyFatStart + ((bodyFatEnd - bodyFatStart) * day) / totalDays) * 10) / 10;
    }
    out.push(point);
  }
  return out;
}

function input(overrides: Partial<DietPhaseInput> = {}): DietPhaseInput {
  return {
    goal: "lose_weight",
    weighIns: series(200, 6, -1.5),
    currentCalories: 2200,
    ...overrides,
  };
}

describe("computeWeightTrend", () => {
  it("smooths daily noise into a trend", () => {
    const noisy = [
      { date: "2026-05-01", weight: 200 },
      { date: "2026-05-02", weight: 204 }, // water spike
      { date: "2026-05-03", weight: 199 },
      { date: "2026-05-04", weight: 200 },
      { date: "2026-05-10", weight: 198 },
      { date: "2026-05-15", weight: 197 },
    ];
    const trend = computeWeightTrend(noisy);

    // The trend must not chase the 204 spike.
    expect(trend.trendWeightLbs).toBeLessThan(201);
    expect(trend.trendWeightLbs).toBeGreaterThan(197);
    expect(trend.latestWeightLbs).toBe(197);
    expect(trend.reliable).toBe(true);
  });

  it("reports weekly rate as both lbs and percent", () => {
    const trend = computeWeightTrend(series(200, 8, -2));
    expect(trend.weeklyChangeLbs).toBeLessThan(0);
    expect(trend.weeklyChangePct).toBeLessThan(0);
    expect(Math.abs(trend.weeklyChangePct)).toBeCloseTo(Math.abs(trend.weeklyChangeLbs) / trend.trendWeightLbs, 3);
  });

  it("is unreliable without enough data", () => {
    expect(computeWeightTrend([]).reliable).toBe(false);
    expect(computeWeightTrend([{ date: "2026-05-01", weight: 200 }]).reliable).toBe(false);
    // Enough weigh-ins but too short a span
    expect(
      computeWeightTrend([
        { date: "2026-05-01", weight: 200 },
        { date: "2026-05-02", weight: 200 },
        { date: "2026-05-03", weight: 199 },
        { date: "2026-05-04", weight: 199 },
      ]).reliable,
    ).toBe(false);
  });

  it("ignores entries without a weight", () => {
    const trend = computeWeightTrend([
      { date: "2026-05-01", weight: 200 },
      { date: "2026-05-05", weight: undefined },
      { date: "2026-05-20", weight: 196 },
    ]);
    expect(trend.weighInCount).toBe(2);
  });
});

describe("computeLeanMassSignal", () => {
  it("flags when too much of the loss is lean mass", () => {
    // 200 → 190 lb, body fat 20% → 19.5%: lean falls ~7 lb of the 10 lb lost.
    const signal = computeLeanMassSignal(series(200, 6, -1.67, 3, 20, 19.5))!;
    expect(signal.leanChangeLbs).toBeLessThan(0);
    expect(signal.leanShareOfLoss).toBeGreaterThan(0.25);
    expect(signal.losingLeanMass).toBe(true);
  });

  it("stays quiet when fat is doing the leaving", () => {
    // 200 → 190 lb, body fat 25% → 20%: lean mass actually rises.
    const signal = computeLeanMassSignal(series(200, 6, -1.67, 3, 25, 20))!;
    expect(signal.losingLeanMass).toBe(false);
  });

  it("returns undefined without body fat readings", () => {
    expect(computeLeanMassSignal(series(200, 6, -1.5))).toBeUndefined();
  });
});

describe("assessDietPhase — cut", () => {
  it("holds steady at a productive rate", () => {
    // ~1.4 lb/week on 200 lb ≈ 0.7%/week
    const result = assessDietPhase(input({ weighIns: series(200, 6, -1.4) }));
    expect(result.phase).toBe("cut");
    expect(result.rateVerdict).toBe("on_track");
    expect(result.calorieAdjustment).toBe(0);
    expect(result.headline).toContain("productive range");
  });

  it("adds calories back when the cut is too aggressive", () => {
    // ~4 lb/week on 200 lb = 2%/week
    const result = assessDietPhase(input({ weighIns: series(200, 6, -4) }));
    expect(result.rateVerdict).toBe("too_fast");
    expect(result.calorieAdjustment).toBeGreaterThan(0);
    expect(result.details.join(" ")).toContain("muscle");
  });

  it("trims calories when loss is too slow", () => {
    const result = assessDietPhase(input({ weighIns: series(200, 6, -0.4) }));
    expect(result.rateVerdict).toBe("too_slow");
    expect(result.calorieAdjustment).toBeLessThan(0);
  });

  it("recognizes a stall and suggests a break after several weeks", () => {
    const result = assessDietPhase(input({ weighIns: series(200, 5, 0) }));
    expect(result.rateVerdict).toBe("stalled");
    expect(result.calorieAdjustment).toBeLessThan(0);
    expect(result.suggestedPhase).toBe("diet_break");
  });

  it("calls a diet break after a long deficit", () => {
    const result = assessDietPhase(
      input({ weighIns: series(200, 6, -1.4), weeksInDeficit: 14, estimatedTDEE: 2700 }),
    );
    expect(result.dietBreakDue).toBe(true);
    expect(result.suggestedPhase).toBe("diet_break");
    expect(result.calorieAdjustment).toBeGreaterThan(0);
    expect(result.headline).toContain("diet break");
  });

  it("overrides an acceptable rate when lean mass is going", () => {
    const result = assessDietPhase(
      input({ weighIns: series(200, 6, -1.4, 3, 20, 19.5) }),
    );
    expect(result.leanMass?.losingLeanMass).toBe(true);
    expect(result.details[0]).toContain("lean mass");
    expect(result.calorieAdjustment).toBeGreaterThan(0);
  });
});

describe("assessDietPhase — lean bulk", () => {
  it("holds at a lean-bulk pace", () => {
    // ~0.7 lb/week on 180 lb ≈ 0.39%/week
    const result = assessDietPhase(
      input({ goal: "build_muscle", weighIns: series(180, 6, 0.7), currentCalories: 3000 }),
    );
    expect(result.phase).toBe("lean_bulk");
    expect(result.rateVerdict).toBe("on_track");
    expect(result.calorieAdjustment).toBe(0);
  });

  it("pulls back when gaining too fast", () => {
    const result = assessDietPhase(
      input({ goal: "build_muscle", weighIns: series(180, 6, 2), currentCalories: 3400 }),
    );
    expect(result.rateVerdict).toBe("too_fast");
    expect(result.calorieAdjustment).toBeLessThan(0);
    expect(result.details.join(" ")).toContain("fat");
  });

  it("adds calories when the scale is flat", () => {
    const result = assessDietPhase(
      input({ goal: "build_muscle", weighIns: series(180, 6, 0), currentCalories: 2800 }),
    );
    expect(result.rateVerdict).toBe("stalled");
    expect(result.calorieAdjustment).toBeGreaterThan(0);
  });

  it("corrects an accidental deficit", () => {
    const result = assessDietPhase(
      input({ goal: "build_muscle", weighIns: series(180, 6, -1), currentCalories: 2600 }),
    );
    expect(result.rateVerdict).toBe("wrong_direction");
    expect(result.calorieAdjustment).toBeGreaterThan(0);
  });
});

describe("assessDietPhase — maintenance", () => {
  it("approves a flat trend", () => {
    const result = assessDietPhase(
      input({ goal: "maintain", weighIns: series(180, 6, 0.05), currentCalories: 2600 }),
    );
    expect(result.phase).toBe("maintenance");
    expect(result.rateVerdict).toBe("on_track");
    expect(result.calorieAdjustment).toBe(0);
  });

  it("corrects drift in either direction", () => {
    const up = assessDietPhase(
      input({ goal: "maintain", weighIns: series(180, 6, 1), currentCalories: 2600 }),
    );
    expect(up.calorieAdjustment).toBeLessThan(0);

    const down = assessDietPhase(
      input({ goal: "maintain", weighIns: series(180, 6, -1), currentCalories: 2600 }),
    );
    expect(down.calorieAdjustment).toBeGreaterThan(0);
  });
});

describe("assessDietPhase — guardrails", () => {
  it("changes nothing when the data is too thin", () => {
    const result = assessDietPhase(input({ weighIns: [{ date: "2026-05-01", weight: 200 }] }));
    expect(result.calorieAdjustment).toBe(0);
    expect(result.headline).toContain("Not enough weigh-ins");
  });

  it("surfaces a drifting TDEE estimate once it is confident", () => {
    const result = assessDietPhase(
      input({ weighIns: series(200, 6, -1.4), currentCalories: 2200, estimatedTDEE: 2800, tdeeConfidence: 70 }),
    );
    expect(result.details.join(" ")).toContain("2800");
  });

  it("stays quiet about TDEE when confidence is low", () => {
    const result = assessDietPhase(
      input({ weighIns: series(200, 6, -1.4), currentCalories: 2200, estimatedTDEE: 2800, tdeeConfidence: 20 }),
    );
    expect(result.details.join(" ")).not.toContain("2800");
  });
});

describe("dietPhaseLabel", () => {
  it("labels every phase", () => {
    expect(dietPhaseLabel("cut")).toBe("Cut");
    expect(dietPhaseLabel("diet_break")).toBe("Diet break");
    expect(dietPhaseLabel("lean_bulk")).toBe("Lean bulk");
    expect(dietPhaseLabel("maintenance")).toBe("Maintenance");
    expect(dietPhaseLabel("recomp")).toBe("Recomp");
  });
});

File: recomp-mobile/app/src/test/java/com/refactor/app/api/DietPhaseTest.kt
package com.refactor.app.api

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.LocalDate

/** Mirrors `src/lib/diet-phase.test.ts` and iOS `DietPhaseTests.swift`. */
class DietPhaseTest {

    /** Weigh-ins every [stepDays] days, changing by [deltaPerWeek] lbs per week. */
    private fun series(
        startWeight: Double,
        weeks: Int,
        deltaPerWeek: Double,
        stepDays: Int = 2,
        bodyFatStart: Double? = null,
        bodyFatEnd: Double? = null,
    ): List<DietPhase.WeighIn> {
        val start = LocalDate.parse("2026-05-01")
        val totalDays = weeks * 7
        val out = mutableListOf<DietPhase.WeighIn>()
        var day = 0
        while (day <= totalDays) {
            val weight = startWeight + (deltaPerWeek * day) / 7
            val bodyFat = if (bodyFatStart != null && bodyFatEnd != null) {
                Math.round((bodyFatStart + ((bodyFatEnd - bodyFatStart) * day) / totalDays) * 10) / 10.0
            } else null
            out.add(
                DietPhase.WeighIn(
                    date = start.plusDays(day.toLong()).toString(),
                    weightLbs = Math.round(weight * 10) / 10.0,
                    bodyFatPercent = bodyFat,
                )
            )
            day += stepDays
        }
        return out
    }

    // ── Trend ───────────────────────────────────────────

    @Test
    fun `smooths daily noise into a trend`() {
        val noisy = listOf(
            DietPhase.WeighIn("2026-05-01", 200.0),
            DietPhase.WeighIn("2026-05-02", 204.0), // water spike
            DietPhase.WeighIn("2026-05-03", 199.0),
            DietPhase.WeighIn("2026-05-04", 200.0),
            DietPhase.WeighIn("2026-05-10", 198.0),
            DietPhase.WeighIn("2026-05-15", 197.0),
        )
        val trend = DietPhase.computeTrend(noisy)

        assertTrue(trend.trendWeightLbs < 201)
        assertTrue(trend.trendWeightLbs > 197)
        assertEquals(197.0, trend.latestWeightLbs, 0.0)
        assertTrue(trend.reliable)
    }

    @Test
    fun `reports weekly rate as both lbs and percent`() {
        val trend = DietPhase.computeTrend(series(200.0, 8, -2.0))
        assertTrue(trend.weeklyChangeLbs < 0)
        assertTrue(trend.weeklyChangePct < 0)
    }

    @Test
    fun `is unreliable without enough data`() {
        assertFalse(DietPhase.computeTrend(emptyList()).reliable)
        assertFalse(DietPhase.computeTrend(listOf(DietPhase.WeighIn("2026-05-01", 200.0))).reliable)
        val shortSpan = listOf(
            DietPhase.WeighIn("2026-05-01", 200.0),
            DietPhase.WeighIn("2026-05-02", 200.0),
            DietPhase.WeighIn("2026-05-03", 199.0),
            DietPhase.WeighIn("2026-05-04", 199.0),
        )
        assertFalse(DietPhase.computeTrend(shortSpan).reliable)
    }

    @Test
    fun `ignores entries without a weight`() {
        val trend = DietPhase.computeTrend(
            listOf(
                DietPhase.WeighIn("2026-05-01", 200.0),
                DietPhase.WeighIn("2026-05-05", null),
                DietPhase.WeighIn("2026-05-20", 196.0),
            )
        )
        assertEquals(2, trend.weighInCount)
    }

    // ── Lean mass ───────────────────────────────────────

    @Test
    fun `flags when too much of the loss is lean mass`() {
        val signal = DietPhase.computeLeanMassSignal(
            series(200.0, 6, -1.67, 3, 20.0, 19.5)
        )!!
        assertTrue(signal.leanChangeLbs < 0)
        assertTrue(signal.leanShareOfLoss > 0.25)
        assertTrue(signal.losingLeanMass)
    }

    @Test
    fun `stays quiet when fat is doing the leaving`() {
        val signal = DietPhase.computeLeanMassSignal(series(200.0, 6, -1.67, 3, 25.0, 20.0))!!
        assertFalse(signal.losingLeanMass)
    }

    @Test
    fun `returns null without body fat readings`() {
        assertNull(DietPhase.computeLeanMassSignal(series(200.0, 6, -1.5)))
    }

    // ── Cut ─────────────────────────────────────────────

    @Test
    fun `holds steady at a productive rate`() {
        val result = DietPhase.assess("lose_weight", series(200.0, 6, -1.4), 2200)
        assertEquals(DietPhase.Name.CUT, result.phase)
        assertEquals(DietPhase.RateVerdict.ON_TRACK, result.rateVerdict)
        assertEquals(0, result.calorieAdjustment)
    }

    @Test
    fun `adds calories back when the cut is too aggressive`() {
        val result = DietPhase.assess("lose_weight", series(200.0, 6, -4.0), 2200)
        assertEquals(DietPhase.RateVerdict.TOO_FAST, result.rateVerdict)
        assertTrue(result.calorieAdjustment > 0)
        assertTrue(result.details.joinToString(" ").contains("muscle"))
    }

    @Test
    fun `trims calories when loss is too slow`() {
        val result = DietPhase.assess("lose_weight", series(200.0, 6, -0.4), 2200)
        assertEquals(DietPhase.RateVerdict.TOO_SLOW, result.rateVerdict)
        assertTrue(result.calorieAdjustment < 0)
    }

    @Test
    fun `recognizes a stall and suggests a break`() {
        val result = DietPhase.assess("lose_weight", series(200.0, 5, 0.0), 2200)
        assertEquals(DietPhase.RateVerdict.STALLED, result.rateVerdict)
        assertTrue(result.calorieAdjustment < 0)
        assertEquals(DietPhase.Name.DIET_BREAK, result.suggestedPhase)
    }

    @Test
    fun `calls a diet break after a long deficit`() {
        val result = DietPhase.assess(
            "lose_weight", series(200.0, 6, -1.4), 2200,
            estimatedTDEE = 2700, weeksInDeficit = 14,
        )
        assertTrue(result.dietBreakDue)
        assertEquals(DietPhase.Name.DIET_BREAK, result.suggestedPhase)
        assertTrue(result.calorieAdjustment > 0)
        assertTrue(result.headline.contains("diet break"))
    }

    @Test
    fun `lean mass loss overrides an acceptable rate`() {
        val result = DietPhase.assess(
            "lose_weight", series(200.0, 6, -1.4, 3, 20.0, 19.5), 2200,
        )
        assertTrue(result.leanMass!!.losingLeanMass)
        assertTrue(result.details.first().contains("lean mass"))
        assertTrue(result.calorieAdjustment > 0)
    }

    // ── Lean bulk ───────────────────────────────────────

    @Test
    fun `holds at a lean-bulk pace`() {
        val result = DietPhase.assess("build_muscle", series(180.0, 6, 0.7), 3000)
        assertEquals(DietPhase.Name.LEAN_BULK, result.phase)
        assertEquals(DietPhase.RateVerdict.ON_TRACK, result.rateVerdict)
        assertEquals(0, result.calorieAdjustment)
    }

    @Test
    fun `pulls back when gaining too fast`() {
        val result = DietPhase.assess("build_muscle", series(180.0, 6, 2.0), 3400)
        assertEquals(DietPhase.RateVerdict.TOO_FAST, result.rateVerdict)
        assertTrue(result.calorieAdjustment < 0)
    }

    @Test
    fun `adds calories when the scale is flat`() {
        val result = DietPhase.assess("build_muscle", series(180.0, 6, 0.0), 2800)
        assertEquals(DietPhase.RateVerdict.STALLED, result.rateVerdict)
        assertTrue(result.calorieAdjustment > 0)
    }

    @Test
    fun `corrects an accidental deficit`() {
        val result = DietPhase.assess("build_muscle", series(180.0, 6, -1.0), 2600)
        assertEquals(DietPhase.RateVerdict.WRONG_DIRECTION, result.rateVerdict)
        assertTrue(result.calorieAdjustment > 0)
    }

    // ── Maintenance ─────────────────────────────────────

    @Test
    fun `approves a flat trend`() {
        val result = DietPhase.assess("maintain", series(180.0, 6, 0.05), 2600)
        assertEquals(DietPhase.Name.MAINTENANCE, result.phase)
        assertEquals(DietPhase.RateVerdict.ON_TRACK, result.rateVerdict)
        assertEquals(0, result.calorieAdjustment)
    }

    @Test
    fun `corrects drift in either direction`() {
        assertTrue(DietPhase.assess("maintain", series(180.0, 6, 1.0), 2600).calorieAdjustment < 0)
        assertTrue(DietPhase.assess("maintain", series(180.0, 6, -1.0), 2600).calorieAdjustment > 0)
    }

    // ── Guardrails ──────────────────────────────────────

    @Test
    fun `changes nothing when the data is too thin`() {
        val result = DietPhase.assess(
            "lose_weight", listOf(DietPhase.WeighIn("2026-05-01", 200.0)), 2200,
        )
        assertEquals(0, result.calorieAdjustment)
        assertTrue(result.headline.contains("Not enough weigh-ins"))
    }

    @Test
    fun `surfaces a drifting TDEE estimate once it is confident`() {
        val result = DietPhase.assess(
            "lose_weight", series(200.0, 6, -1.4), 2200,
            estimatedTDEE = 2800, tdeeConfidence = 70,
        )
        assertTrue(result.details.joinToString(" ").contains("2800"))
    }

    @Test
    fun `stays quiet about TDEE when confidence is low`() {
        val result = DietPhase.assess(
            "lose_weight", series(200.0, 6, -1.4), 2200,
            estimatedTDEE = 2800, tdeeConfidence = 20,
        )
        assertFalse(result.details.joinToString(" ").contains("2800"))
    }

    @Test
    fun `phase labels`() {
        assertEquals("Cut", DietPhase.Name.CUT.label)
        assertEquals("Diet break", DietPhase.Name.DIET_BREAK.label)
        assertEquals("Lean bulk", DietPhase.Name.LEAN_BULK.label)
        assertEquals("Maintenance", DietPhase.Name.MAINTENANCE.label)
        assertEquals("Recomp", DietPhase.Name.RECOMP.label)
    }
}
